src/coreclr/minipal/Windows/doublemapping.cpp

```diff
@@ -0,0 +1,246 @@
+
+    if (size == 0)
+    {
+        return nullptr;
+    }
+
+    //
+    // First lets normalize the pMinAddr and pMaxAddr values
+    //
+    // If pMinAddr is NULL then set it to BOT_MEMORY
+    if ((pMinAddr == 0) || (pMinAddr < (BYTE *) BOT_MEMORY))
+    {
+        pMinAddr = (BYTE *) BOT_MEMORY;
+    }
+
+    // If pMaxAddr is NULL then set it to TOP_MEMORY
+    if ((pMaxAddr == 0) || (pMaxAddr > (BYTE *) TOP_MEMORY))
+    {
+        pMaxAddr = (BYTE *) TOP_MEMORY;
+    }
+
+    // If pMaxAddr is not greater than pMinAddr we can not make an allocation
+    if (pMaxAddr <= pMinAddr)
+    {
+        return nullptr;
+    }
+
+    // If pMinAddr is BOT_MEMORY and pMaxAddr is TOP_MEMORY
+    // then we can call ClrVirtualAlloc instead
+    if ((pMinAddr == (BYTE *) BOT_MEMORY) && (pMaxAddr == (BYTE *) TOP_MEMORY))
+    {
+        return (BYTE*)MapViewOfFile((HANDLE)mapperHandle,
+                        FILE_MAP_EXECUTE | FILE_MAP_READ | FILE_MAP_WRITE,
+                        HIDWORD((int64_t)offset),
+                        LODWORD((int64_t)offset),
+                        size);
+    }
+
+    // We will do one scan from [pMinAddr .. pMaxAddr]
+    // First align the tryAddr up to next 64k base address.
+    // See docs for VirtualAllocEx and lpAddress and 64k alignment for reasons.
+    //
+    BYTE *   tryAddr            = (BYTE *)ALIGN_UP((BYTE *)pMinAddr, VIRTUAL_ALLOC_RESERVE_GRANULARITY);
+    bool     virtualQueryFailed = false;
+    bool     faultInjected      = false;
+    unsigned virtualQueryCount  = 0;
+
+    // Now scan memory and try to find a free block of the size requested.
+    while ((tryAddr + size) <= (BYTE *) pMaxAddr)
+    {
+        MEMORY_BASIC_INFORMATION mbInfo;
+
+        // Use VirtualQuery to find out if this address is MEM_FREE
+        //
+        virtualQueryCount++;
+        if (!VirtualQuery((LPCVOID)tryAddr, &mbInfo, sizeof(mbInfo)))
+        {
+            // Exit and return nullptr if the VirtualQuery call fails.
+            virtualQueryFailed = true;
+            break;
+        }
+
+        // Is there enough memory free from this start location?
+        // Note that for most versions of UNIX the mbInfo.RegionSize returned will always be 0
+        if ((mbInfo.State == MEM_FREE) &&
+            (mbInfo.RegionSize >= (SIZE_T) size || mbInfo.RegionSize == 0))
+        {
+            // Try reserving the memory using VirtualAlloc now
+            pResult = (BYTE*)MapViewOfFileEx((HANDLE)mapperHandle,
+                        FILE_MAP_EXECUTE | FILE_MAP_READ | FILE_MAP_WRITE,
+                        HIDWORD((int64_t)offset),
+                        LODWORD((int64_t)offset),
+                        size,
+                        tryAddr);
+
+            // Normally this will be successful
+            //
+            if (pResult != nullptr)
+            {
+                // return pResult
+                break;
+            }
+
+#ifdef _DEBUG
+            // if (ShouldInjectFaultInRange())
```

Comment:
@jkotas do you think it is worth preserving this fault injection? I would need to either read the env var directly in this minipal or pass its state down from the executable allocator.

Reply:
We do not need to. The fault injection was not ever used in .NET Core as far as I know.